mijia/src/lib.rs

```diff
@@ -1,137 +1,104 @@
-    BluetoothSession,
-};
+use bluez_generated::generated::gattcharacteristic1::OrgBluezGattCharacteristic1;
+use dbus::arg::RefArg;
+use dbus::nonblock::stdintf::org_freedesktop_dbus::ObjectManager;
+use dbus::nonblock::SyncConnection;
+use itertools::Itertools;
 use std::cmp::max;
 use std::collections::HashMap;
 use std::convert::TryInto;
-use std::error::Error;
 use std::fmt::{Display, Formatter};
 use std::fs::File;
 use std::io::{self, BufRead, BufReader, ErrorKind};
-use std::thread;
+use std::sync::Arc;
 use std::time::Duration;
 
-const SCAN_DURATION: Duration = Duration::from_millis(5000);
-const CONNECT_TIMEOUT_MS: i32 = 10_000;
-
-const MIJIA_SERVICE_DATA_UUID: &str = "0000fe95-0000-1000-8000-00805f9b34fb";
+pub const MIJIA_SERVICE_DATA_UUID: &str = "0000fe95-0000-1000-8000-00805f9b34fb";
 pub const SERVICE_CHARACTERISTIC_PATH: &str = "/service0021/char0035";
-const CONNECTION_INTERVAL_CHARACTERISTIC_PATH: &str = "/service0021/char0045";
+pub const CONNECTION_INTERVAL_CHARACTERISTIC_PATH: &str = "/service0021/char0045";
 /// 500 in little-endian
-const CONNECTION_INTERVAL_500_MS: [u8; 3] = [0xF4, 0x01, 0x00];
-
-pub fn scan(bt_session: &BluetoothSession) -> Result<Vec<String>, Box<dyn Error>> {
-    let adapter: BluetoothAdapter = BluetoothAdapter::init(bt_session)?;
-    adapter.set_powered(true)?;
-
-    let discover_session =
-        BluetoothDiscoverySession::create_session(&bt_session, adapter.get_id())?;
-    discover_session.start_discovery()?;
-    println!("Scanning");
-    // Wait for the adapter to scan for a while.
-    thread::sleep(SCAN_DURATION);
-    let device_list = adapter.get_device_list()?;
-
-    discover_session.stop_discovery()?;
+pub const CONNECTION_INTERVAL_500_MS: [u8; 3] = [0xF4, 0x01, 0x00];
 
-    println!("{:?} devices found", device_list.len());
-
-    Ok(device_list)
+pub struct SensorProps {
+    pub object_path: String,
+    pub mac_address: String,
 }
 
-pub fn find_sensors<'a>(
-    bt_session: &'a BluetoothSession,
-    device_list: &[String],
-) -> Vec<BluetoothDevice<'a>> {
-    let mut sensors = vec![];
-    for device_path in device_list {
-        let device = BluetoothDevice::new(bt_session, device_path.to_string());
-        println!(
-            "Device: {:?} Name: {:?}",
-            device_path,
-            device.get_name().ok()
-        );
-        if let Ok(service_data) = device.get_service_data() {
-            println!("Service data: {:?}", service_data);
+pub async fn get_sensors(
+    dbus_conn: Arc<SyncConnection>,
+) -> Result<Vec<SensorProps>, anyhow::Error> {
+    let bluez_root =
+        dbus::nonblock::Proxy::new("org.bluez", "/", Duration::from_secs(30), dbus_conn.clone());
+    let tree = bluez_root.get_managed_objects().await?;
+
+    let paths = tree
+        .into_iter()
+        .filter_map(|(path, interfaces)| {
+            // FIXME: can we generate a strongly typed deserialiser for this,
+            // based on the introspection data?
```

Comment:
- [ ] add a ticket for this when we make a new repo (what should the new repo be called?)